### PR TITLE
Moth Jump Finished

### PR DESCRIPTION
**Pull request by aaronspringer**
## The dip equation
image

### CODE REVIEW

File: scripts/entity/components/movement_components/player_movement_component.gd [hunk 1]
```diff
@@ -4,58 +4,96 @@ class_name PlayerMovementComponent extends BaseMovementComponent
...
+				# Maintain the current height
+				velocity.y = 0.0
+				_moth_jump_timer += delta
+			else:
+				# End Moth Jump
+				_is_moth_jumping = false
+				velocity.y += super.get_gravity().y * delta  # Resume normal gravity
+	else:
+		# Apply normal gravity
+		velocity.y += super.get_gravity().y * delta
```
Comment (SllimRevilo) [inline scripts/entity/components/movement_components/player_movement_component.gd]: this one line seems like the function apply_gravity

File: scripts/entity/components/movement_components/player_movement_component.gd [hunk 2]
```diff
@@ -4,58 +4,96 @@ class_name PlayerMovementComponent extends BaseMovementComponent
...
 		on_jump_started.emit()
-
+	elif not _is_moth_jumping:
+		# Start Moth Jump
+		_is_moth_jumping = true
+		_is_fluttering = true
+		_moth_jump_timer = 0.0
+		_moth_jump_flutter_timer = 0.0
+		_moth_jump_initial_height = global_position.y
+		_is_jump_button_held = true
```
Comment (SllimRevilo) [inline scripts/entity/components/movement_components/player_movement_component.gd]: Id put this in like an init_moth_jump_function()

Comment (SllimRevilo) [reply scripts/entity/components/movement_components/player_movement_component.gd]: We might also want like a on_moth_jump signal or something like that also

File: scripts/entity/components/movement_components/player_movement_component.gd [hunk 3]
```diff
@@ -4,58 +4,96 @@ class_name PlayerMovementComponent extends BaseMovementComponent
...
+				velocity.y = 0.0
+		else:
+			if _is_jump_button_held and _moth_jump_timer < MOTH_JUMP_HANGTIME:
+				# Maintain the current height
+				velocity.y = 0.0
+				_moth_jump_timer += delta
+			else:
+				# End Moth Jump
+				_is_moth_jumping = false
+				velocity.y += super.get_gravity().y * delta  # Resume normal gravity
```
Comment (SllimRevilo) [inline scripts/entity/components/movement_components/player_movement_component.gd]: Id probably put this into a function or two and explain what you are doing. There is a lot of math but I cant tell you what goes to what easily.

File: scripts/entity/components/movement_components/player_movement_component.gd [hunk 4]
```diff
@@ -4,58 +4,96 @@ class_name PlayerMovementComponent extends BaseMovementComponent
...
+			var flutter_progress: float = _moth_jump_flutter_timer / MOTH_JUMP_FLUTTER_DURATION
+			# Copy in to desmos to see why I chose this function: -\sin\left(\frac{\left(x-a\right)^{2}}{c}\right)+1
+			var flutter_dip: float = -MOTH_JUMP_DIP_AMOUNT * (-sin((pow((flutter_progress-0.5),2))/0.16) + 1)
+			var desired_position_y: float = _moth_jump_initial_height + flutter_dip
+			var position_difference_y: float = desired_position_y - global_position.y
+			velocity.y = position_difference_y / delta
+			_moth_jump_flutter_timer += delta
+			if _moth_jump_flutter_timer >= MOTH_JUMP_FLUTTER_DURATION:
+				_is_fluttering = false
+				global_position.y = _moth_jump_initial_height
```
Comment (SllimRevilo) [inline scripts/entity/components/movement_components/player_movement_component.gd]: Why do you have to set the position if it should already get there? The code most likely is just redundant

File: scripts/entity/components/movement_components/player_movement_component.gd [hunk 5]
```diff
@@ -4,58 +4,96 @@ class_name PlayerMovementComponent extends BaseMovementComponent
 ## The max horizontal speed the entity can move.
 @export var MAX_SPEED: int = 200
 ## The jump velocity of the entity.
-@export var JUMP_VELOCITY: int = -300
+@export var JUMP_VELOCITY: int = -350
```
Comment (Smoresies) [inline scripts/entity/components/movement_components/player_movement_component.gd]: This should just be a positive number, and made negative in the math it's used in

File: scripts/entity/components/movement_components/player_movement_component.gd [hunk 6]
```diff
@@ -4,58 +4,96 @@ class_name PlayerMovementComponent extends BaseMovementComponent
...
 ## The horizontal movement acceleration for the entity.
 @export var ACCELERATION: int = 1500
 ## The time it takes to fully decelerate from horizontal input in seconds.
 @export_range(0, 10, 0.1, "or_greater") var DECELERATION_TIME: float = 0.2
 
+## Moth Jump variables
+@export var MOTH_JUMP_HANGTIME: float = 1.0
+@export var MOTH_JUMP_DIP_AMOUNT: float = -18.0
```
Comment (Smoresies) [inline scripts/entity/components/movement_components/player_movement_component.gd]: Also should just be positive and adjusted in actual code sections/updated in _ready to be negative. Easier on designers to deal with positive numbers and not need to worry about the fact it needs to be negative in the background.

File: scripts/entity/components/movement_components/player_movement_component.gd (hunk 2 above)
Comment (Smoresies) [reply scripts/entity/components/movement_components/player_movement_component.gd]: Seconded

File: scripts/entity/components/movement_components/player_movement_component.gd [hunk 7]
```diff
@@ -4,58 +4,96 @@ class_name PlayerMovementComponent extends BaseMovementComponent
...
-	# do nothing
-	pass
-
+	_is_jump_button_held = false
 
 ## Moves the entity based on the current input direction.
 func apply_horizontal_movement(delta: float) -> void:
- 	# Direction button active
-	if _current_input_direction:
+	if _current_input_direction != 0:
```
Comment (Smoresies) [inline scripts/entity/components/movement_components/player_movement_component.gd:68]: Why the change? Personally felt the original was more readable.

Comment (SllimRevilo) [reply scripts/entity/components/movement_components/player_movement_component.gd:68]: I personally prefer this since it is more explicit in what it is doing and leaves less room for interpretation. I think either way is fine though.

File: scripts/entity/components/movement_components/player_movement_component.gd (hunk 6 above)
Comment (aaronspringer) [reply scripts/entity/components/movement_components/player_movement_component.gd]: Used _ready to invert

File: scripts/entity/components/movement_components/player_movement_component.gd (hunk 5 above)
Comment (aaronspringer) [reply scripts/entity/components/movement_components/player_movement_component.gd]: Used _ready to invert

File: scripts/entity/components/movement_components/player_movement_component.gd (hunk 4 above)
Comment (aaronspringer) [reply scripts/entity/components/movement_components/player_movement_component.gd]: It was, deleted.

File: scripts/entity/components/movement_components/player_movement_component.gd (hunk 2 above)
Comment (aaronspringer) [reply scripts/entity/components/movement_components/player_movement_component.gd]: Created signals. Moved in to _init_moth_jump()

File: scripts/entity/components/movement_components/player_movement_component.gd (hunk 1 above)
Comment (aaronspringer) [reply scripts/entity/components/movement_components/player_movement_component.gd]: Used super.apply_gravity() for the default gravity function.

File: scripts/entity/components/movement_components/player_movement_component.gd (hunk 7 above)
Comment (aaronspringer) [reply scripts/entity/components/movement_components/player_movement_component.gd:68]: Yeah for verbosity, I don't really care though, whichever works

File: scripts/entity/components/movement_components/player_movement_component.gd [hunk 8]
```diff
@@ -1,61 +1,134 @@
...
+
+			else:
+				# End Moth Jump
+				_is_moth_jumping = false
+				on_moth_jump_ended.emit()
+				super.apply_gravity(delta)  # Resume normal gravity
+
+	else:
+		# Apply normal gravity
+		super.apply_gravity(delta)
```
Comment (Smoresies) [inline scripts/entity/components/movement_components/player_movement_component.gd]: Small organization concern, but it'd be better to early exit and have less nesting. Video example posting in programming section of discord (https://www.youtube.com/watch?v=CFRhGnuXG-4&pp=ygUMbmV2ZXIgbmVzdGVy)

Comment (aaronspringer) [reply scripts/entity/components/movement_components/player_movement_component.gd]: Nested less

Comment (SllimRevilo) [reply scripts/entity/components/movement_components/player_movement_component.gd]: Id just be concerned mostly about readability with it

File: scripts/entity/components/movement_components/player_movement_component.gd (hunk 3 above)
Comment (aaronspringer) [reply scripts/entity/components/movement_components/player_movement_component.gd]: Refactored

**Review by SllimRevilo — APPROVED** [review]
